compiler/typecheck/TcDaml.hs

```diff
@@ -13,34 +11,310 @@ import OccName
+               (map (ppr . fst) ifaces)
+           ]
+
+data DamlVariant = Template Name | Interface Name | Choice Name
+
+data DamlInfo = DamlInfo
+  { templates :: [Name]
+  , interfaces :: [Name]
+  , choices :: [(Name, Name)]
+  , methods :: [(FastString, (Name, Type))]
+  , implementations :: [(Name, Name)]
+  , views :: [(Name, Type)]
+  }
+
+dedupe :: DamlInfo -> DamlInfo
+dedupe (DamlInfo x0 x1 x2 x3 x4 x5) =
+  DamlInfo
+    (nubSort x0)
+    (nubSort x1)
+    (nubSort x2)
+    (nubSortBy (\(mName, (cName, type_)) -> (mName, cName)) x3)
+    (nubSort x4)
+    (nubSortBy fst x5)
+  where
+    nubSortBy :: Ord b => (a -> b) -> [a] -> [a]
+    nubSortBy f xs = M.elems $ M.fromList $ map (\x -> (f x, x)) xs
+
+isTemplate info name = name `elem` templates info
+isInterface info name = name `elem` interfaces info
+variantName info name
+  | isTemplate info name = text "template" <+> ppr name
+  | isInterface info name = text "interface" <+> ppr name
+  | otherwise = text "type" <+> ppr name
+allImplementedInterfaces info name = [iface | (tpl, iface) <- implementations info, name == tpl]
+allImplementingTemplates info name = [tpl | (tpl, iface) <- implementations info, name == iface]
+implements info tpl iface = (tpl, iface) `elem` implementations info
+choiceImplementor info name = [tplOrIface | (tplOrIface, choice) <- choices info, name == choice]
+interfaceView info name = lookup name (views info)
+definesMethod info method = [nameAndType | (m, nameAndType) <- methods info, m == method]
+
+instance Monoid DamlInfo where
+  mempty = DamlInfo [] [] [] [] [] []
+
+instance Semigroup DamlInfo where
+  (<>) (DamlInfo a0 a1 a2 a3 a4 a5) (DamlInfo b0 b1 b2 b3 b4 b5) =
+    DamlInfo (a0 P.<> b0) (a1 P.<> b1) (a2 P.<> b2) (a3 P.<> b3) (a4 P.<> b4) (a5 P.<> b5)
+
+instance Outputable DamlInfo where
+  ppr info =
+    hang (text "DamlInfo {") 2 $
+    vcat [ text "templates:" <+> ppr (templates info)
+         , text "interfaces:" <+> ppr (interfaces info)
+         , text "choices:" <+> ppr (choices info)
+         , text "methods:" <+> ppr (methods info)
+         , text "implementations:" <+> ppr (implementations info)
+         , text "views:" <+> ppr (views info)
+         ]
+
+extractDamlInfoFromGblEnv :: TcGblEnv -> DamlInfo
+extractDamlInfoFromGblEnv env =
+  foldMap extractDamlInfoFromTyThing (typeEnvElts $ tcg_type_env env) `mappend`
+    foldMap extractDamlInfoFromClsInst (instEnvElts $ tcg_inst_env env)
+
+extractDamlInfoFromEPS :: ExternalPackageState -> DamlInfo
+extractDamlInfoFromEPS eps =
+  foldMap extractDamlInfoFromTyThing (typeEnvElts $ eps_PTE eps) `mappend`
+    foldMap extractDamlInfoFromClsInst (instEnvElts $ eps_inst_env eps)
+
+extractDamlInfoFromTyThing :: TyThing -> DamlInfo
+extractDamlInfoFromTyThing tything =
+  case tything of
+    AnId id ->
+      mempty
+        { methods = mapMaybe matchMethod [id]
+        , implementations = mapMaybe matchImplements [id]
+        , choices = mapMaybe matchChoice [id]
```

Comment:
As far as I could tell these are also found by `extractDamlInfoFromClsInst`, so we should be able to remove these without losing any information. Did you find any cases during your tests where only `extractDamlInfoFromTyThing` picked one of these up?

Reply:
I didn't check explicitly, I wrote those lines before I had ClsInst working - I'd rather keep it around, just in the off-chance that it finds something.

Reply:
Nevermind, I'm removing them - they *do* represent a maintenance burden

Reply:
interesting! please let me know your findings

Reply:
Looks like I was wrong about being wrong, I've removed the AnId check - the new debugging output makes it easier to see where difference inferences come from